### DISCUSSION TITLE
Setting bcrypt salt rounds via emqx helm chart additional config values

### DISCUSSION THREAD

**Discussion by PhiRie** [emqx/emqx#12288] [Q&A] [answered]
Hi, 

In our cluster we use bcrypt hash algorithm and I tried to set the salt rounds explicitly via helm additional config parameters like: 

`EMQX_AUTHENTICATION__1__PASSWORD_HASH_ALGORITHM__SALT_ROUNDS: 10`

If i redeploy the helm chart i always get the following error:

`2024-01-10T08:30:11.871664+00:00 [error] #{kind => validation_error,matched_type => "authn:postgresql/authn_hash:bcrypt",path => "authentication.1.password_hash_algorithm",reason => unknown_fields,unknown => "salt_rounds",unmatched => none}
ERROR: call_hocon_failed: -v -t 2024.01.10.08.30.03 -s emqx_conf_schema -c /opt/emqx/etc/emqx.conf -d /opt/emqx/data/configs generate`

So it seems that emqx don't accepts the configuration parameter salt_rounds. Is there anything wrong with the syntax I use because this works for other parameters like the algorithm name itself (EMQX_AUTHENTICATION__1__PASSWORD_HASH_ALGORITHM__NAME: bcrypt). I'm referencing to the documentation here: 

https://www.emqx.io/docs/en/v5.4.1/hocon/#V-authentication-S-builtin_db-password_hash_algorithm-S-bcrypt_rw-salt_rounds
https://github.com/emqx/emqx-docker?tab=readme-ov-file (see configuration for translation of additional config parameters into emqx.conf)

PS: We are on latest version 5.4.1 and I know that the default value for the salt rounds is 10 but for me it would be good to set it explicitly in the config so that other developers also know what is used. Further I also know that there is a limitation since version 5.2.1: https://github.com/emqx/emqx/releases/tag/v5.2.1

**Comment by ieQu1** [comment, marked as answer]
Hello,

There are two classes of authentication backends in EMQX: read-only and read-write. `salt_rounds` is a property applicable to the read-write backends only. Currently we have only one read-write backend: "builtin" (based on Mria database). But it looks like you're using Postgres; it's a read-only backend, hence the error.

**Comment by PhiRie** [comment]
Thx @ieQu1 for your response and clarification. I now also found the difference between builtin database and postgres in the documentation here: 

[Builtin_DB](https://www.emqx.io/docs/en/v5.4.1/hocon/#T-authentication-S-builtin_db)
[Postgres](https://www.emqx.io/docs/en/v5.4.1/hocon/#T-authentication-S-postgresql)
